create_layout.py

```diff
@@ -91,48 +91,123 @@
 import in_toto.models.link
 import in_toto.models.layout
 
-def create_material_rules(links, index):
-  """Create generic material rules (3 variants)
+def changes_between_snapshots(before_dict, after_dict):
+  """Given two 'snapshots' of an artifacts structure -- 'before' and 'after' --
+  return a tuple specifying which artifacts have been added, which have been
+  removed, which have been modified, and which have remained unchanged. Both
+  these dictionaries have artifact names as the keys and their hashes as the
+  values."""
+
+  before_set = set(before_dict.keys())
+  after_set = set(after_dict.keys())
+
+  removed_artifacts = before_set.difference(after_set)
+  added_artifacts = after_set.difference(before_set)
+
+  unchanged_artifacts = set()
+  modified_artifacts = set()
+  for key in before_set.intersection(after_set):
+    if before_dict[key] == after_dict[key]:
+      unchanged_artifacts.add(key)
+    else:
+      modified_artifacts.add(key)
+
+  return (sorted(unchanged_artifacts), sorted(modified_artifacts), sorted(added_artifacts),
+      sorted(removed_artifacts))
```

Comment:
Is it necessary to convert the sets to ordered lists here? Don't you convert them back to sets in the calling functions anyway?

To solve the determinism issue we discussed, I think it would be easier to just do: `for artifact in sorted(...): ...` in all the for-loops in your `create_*_rules` functions